snakemake/dag.py

```diff
@@ -1257,30 +1257,39 @@ async def update_needrun(job):
                             # The first pass (with depends_on_checkpoint_target == True) is not informative
                             # for determining any other changes than file modification dates, as it will
                             # change after evaluating the input function of the job in the second pass.
-                            if RerunTrigger.PARAMS in self.workflow.rerun_triggers:
-                                reason.params_changed = any(
-                                    self.workflow.persistence.params_changed(job)
-                                )
-                            if RerunTrigger.INPUT in self.workflow.rerun_triggers:
-                                reason.input_changed = any(
-                                    self.workflow.persistence.input_changed(job)
-                                )
-                            if RerunTrigger.CODE in self.workflow.rerun_triggers:
-                                reason.code_changed = any(
-                                    [
-                                        f
-                                        async for f in job.outputs_older_than_script_or_notebook()
-                                    ]
-                                ) or any(self.workflow.persistence.code_changed(job))
-                            if (
-                                RerunTrigger.SOFTWARE_ENV
-                                in self.workflow.rerun_triggers
-                            ):
-                                reason.software_stack_changed = any(
-                                    self.workflow.persistence.conda_env_changed(job)
-                                ) or any(
-                                    self.workflow.persistence.container_changed(job)
-                                )
+
+                            if not self.workflow.persistence.has_metadata(job):
+                                reason.no_metadata = True
+                            else:
```

Comment:
This change causes a bug: when using `--drop-metadata` (for filesystem speed), changes except file mtime (especially mtime of `script`/`notebook`) no longer trigger the workflow. I think we should handle this case, or make `no_metadata` a choice of RerunTrigger.

Reply:
Proposed a fix in #3148